stub-connector/src/test/java/uk/gov/ida/notification/apprule/SamlRequestValidatorAppRuleTests.java

```diff
@@ -0,0 +1,156 @@
+        String html = getEidasRequest();
+
+        Document document = Jsoup.parse(html);
+        Element samlRequest = document.select("input[name=\"SAMLRequest\"]").first();
+        return Base64.decodeAsString(samlRequest.val());
+    }
+
+    private String getAuthnRequestIdFromSession() throws URISyntaxException {
+        String decodedSaml = getRequestSaml();
+
+        Document saml = Jsoup.parse(decodedSaml, "", Parser.xmlParser());
+        Element entityDescriptor = saml.select("saml2p|AuthnRequest").first();
+        return entityDescriptor.attr("ID");
+    }
+
+    private String getValidity(String samlMessage) throws URISyntaxException {
+        String html = postEidasResponse(samlMessage);
+
+        Document responseDoc = Jsoup.parse(html);
+        Element validityDiv = responseDoc.select("h2").first();
+
+        String result = validityDiv.text();
+
+        String subStringForValidity = "Saml Validity: ";
+        if (result.contains(subStringForValidity))
+            result = result.replace(subStringForValidity, "");
+
+        return result;
+    }
+
+    private Response eidasSamlMessage(String authid) {
+
+        Attribute firstname = new EidasAttributeBuilder(
+                AttributeConstants.EIDAS_CURRENT_GIVEN_NAME_ATTRIBUTE_NAME,
+                AttributeConstants.EIDAS_CURRENT_GIVEN_NAME_ATTRIBUTE_FRIENDLY_NAME,
+                CurrentGivenNameType.TYPE_NAME,
+                "Jazzy Harrold"
+        ).build();
+
+        ArrayList<Attribute> eidasAttributes = new ArrayList<>();
+
+        eidasAttributes.add(firstname);
+
+        DateTime now = DateTime.now();
+
+        Response response = EidasResponseBuilder.createEidasResponse(
+                "http://stub-connector/",
+                StatusCode.SUCCESS,
+                UUID.randomUUID().toString(),
+                EidasConstants.EIDAS_LOA_SUBSTANTIAL,
+                eidasAttributes,
+                authid,
+                now,
+                now,
+                now,
+                "http://stub-connector/SAML2/Response/POST",
+                "http://localhost:5000/Metadata"
+        );
+
+        return response;
+    }
+
+    private Response signResponse(Response response) {
+        Credential signingCredential = new TestCredentialFactory(
+                TEST_RP_PUBLIC_SIGNING_CERT,
+                TEST_RP_PRIVATE_SIGNING_KEY
+        ).getSigningCredential();
+
+        SamlObjectSigner signer = new SamlObjectSigner(
+                signingCredential.getPublicKey(),
+                signingCredential.getPrivateKey(),
+                TEST_RP_PUBLIC_SIGNING_CERT
+        );
+
+        signer.sign(response);
+
+        return response;
+    }
+
+    private String responseToString(Response response) {
+        SamlObjectMarshaller marshaller = new SamlObjectMarshaller();
+        return marshaller.transformToString(response);
+    }
+
+    @Test
+    public void validSamlRequest() throws Exception {
```

Comment:
Can we move the tests above the private methods?

Reply:
Done